### PR TITLE
feat(storage): add handling of non-ascii object metadata values

### PR DESCRIPTION
**Pull request by HuiSF**
*Issue #, if available:*

*Description of changes:*

1. Renamed `S3GetUrlPluginOptions` `checkObjectExistence` property as `validateObjectExistence`
2. enabled metadata headers in the `infra` storage stack
3. Added metadata non-ascii value encoding and decoding following [S3 recommendation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingMetadata.html) and integration tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/storage/amplify_storage_s3/example/integration_test/use_case_test.dart [hunk 1]
```diff
@@ -472,6 +463,57 @@ void main() {
...
+                isNotEmpty,
+              ),
+            ),
+          );
+
+          final metadata = result.uploadedItem.metadata;
+
+          expect(metadata, equals(testMetadata));
+
+          await Future<void>.delayed(const Duration(seconds: 30));
```
Comment (dnys1) [inline packages/storage/amplify_storage_s3/example/integration_test/use_case_test.dart:513]: Why is this needed?

File: packages/storage/amplify_storage_s3_dart/lib/src/model/s3_item.dart [hunk 2]
```diff
@@ -76,16 +75,32 @@ class S3Item extends StorageItem {
...
+  @internal
+  factory S3Item.fromGetObjectOutput(
+    s3.GetObjectOutput getObjectOutput, {
+    required String key,
+  }) {
+    return S3Item(
+      key: key,
+      lastModified: getObjectOutput.lastModified,
+      eTag: getObjectOutput.eTag,
+      metadata: decodeMetadata(getObjectOutput.metadata?.toMap() ?? const {}),
```
Comment (dnys1) [inline packages/storage/amplify_storage_s3_dart/lib/src/model/s3_item.dart:96]: I would make this part of the `S3Item` constructor. Seems like it's easy to miss this but required on every instantiation.

File: packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/task/s3_upload_task.dart [hunk 3]
```diff
@@ -318,7 +318,8 @@ class S3UploadTask {
         ..body = body
         ..contentType = body.contentType ?? fallbackContentType
         ..key = _resolvedKey
-        ..metadata.addAll(_s3PluginOptions.metadata ?? const {});
+        ..metadata
+            .addAll(encodeMetadata(_s3PluginOptions.metadata ?? const {}));
```
Comment (dnys1) [inline packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/task/s3_upload_task.dart:322]: I would do the following on the plugin options:

```dart
final Map<String, String> _metadata;
Map<String, String> get metadata => encodeMetadata(_metadata);
```

File: packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/utils.dart [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+/// Decode values of [metadata].
+///
+/// A value that is a utf8 encoded string will be decoded.
+Map<String, String> decodeMetadata(Map<String, String> metadata) {
+  return Map.fromEntries(
+    metadata.entries.map((e) => MapEntry(e.key, _rfc2047UTF8Decode(e.value))),
+  );
+}
+
+bool _shouldEncode(String str) {
```
Comment (dnys1) [inline packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/utils.dart:25]:
```suggestion
bool _isUtf8(String str) {
```

File: packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/utils.dart [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  if (match != null) {
+    final encodedStr = match.group(1);
+    if (encodedStr != null) {
+      return utf8.decode(base64.decode(encodedStr));
+    }
+  }
+
+  return str;
+}
```
Comment (dnys1) [inline packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/utils.dart:50]:
```suggestion
final _regexp = RegExp(r'^\?UTF-8\?B\?(.*)\?=$');
String _rfc2047UTF8Decode(String str) {
  final encodedStr = _regexp.firstMatch(str)?.group(1);
  if (encodedStr == null) {
    return str;
  }
  return utf8.decode(base64.decode(encodedStr));
}
```

**Review by dnys1 — COMMENTED** [review]
Could be missing it - do you have tests for `validateObjectExistence: true` where the object doesn't exist?

File: packages/storage/amplify_storage_s3/example/integration_test/use_case_test.dart (hunk 1 above)
Comment (HuiSF) [reply packages/storage/amplify_storage_s3/example/integration_test/use_case_test.dart:513]: Oops debugging code got committed.

File: packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/utils.dart (hunk 4 above)
Comment (HuiSF) [reply packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/utils.dart]: Agreed, changed the function name to `_hasNonAscii`.

File: packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/utils.dart [hunk 6]
```diff
@@ -0,0 +1,48 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import 'dart:convert';
```
Comment (dnys1) [inline packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/utils.dart:4]: Can you mark this library `@internal`?

**Review by dnys1 — APPROVED** [review]

**Review by NikaHsn — APPROVED** [review]

File: packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/utils.dart (hunk 6 above)
Comment (HuiSF) [reply packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/utils.dart:5]: Added 👍
